src/hooks/useLocalStorage.ts

```diff
@@ -0,0 +1,72 @@
+// Retrieved and modified from https://usehooks-ts.com/react-hook/use-local-storage
+
+import { useCallback, useEffect, useState } from 'react'
+import { LOCAL_STORAGE_EVENT } from '~/constants/localStorage'
+
+export const useLocalStorage = <T>(
+  key: string | null,
+  initialValue?: T
+): readonly [T | undefined, (value?: T) => void] => {
+  // Get from local storage then
+  // parse stored json or return initialValue
+  const readValue = useCallback(() => {
+    // Prevent build error "window is undefined" but keep keep working
+    if (typeof window === 'undefined') {
+      return initialValue
+    }
+    if (!key) {
+      return
+    }
+    try {
+      const item = window.localStorage.getItem(key)
+      return item ? JSON.parse(item) : initialValue
+    } catch (error) {
+      return initialValue
+    }
+  }, [initialValue, key])
+  // State to store our value
+  // Pass initial state function to useState so logic is only executed once
+  const [storedValue, setStoredValue] = useState(readValue)
+  // Return a wrapped version of useState's setter function that ...
+  // ... persists the new value to localStorage.
+  const setValue = (value?: T) => {
```

Comment:
I think we need something like this instead if we want the exact same API as `useState`.
```suggestion
  const setValue = (value?: React.Dispatch<React.SetStateAction<T>>) => {
```